### PR TITLE
Feature provider delegation

### PR DESCRIPTION
**Pull request by drmaas**
Refactoring openstack4j provider to be split amongst the various sub-api versions. This will help when migrating api versions, and make the client code and tests easier to maintain.

I also added in a feature to reauthenticate the client if the token is expired, so Spinnaker can remain up for long periods of time.

### CODE REVIEW

File: clouddriver-openstack/src/main/groovy/com/netflix/spinnaker/clouddriver/openstack/client/OpenstackClientProvider.groovy [hunk 1]
```diff
@@ -16,914 +16,36 @@
...
-   */
-  void updateStack(String region, String stackName, String stackId, String template, Map<String, String> subtemplate, ServerGroupParameters parameters) {
-    handleRequest {
-      Map<String, String> params = parameters.toParamsMap()
-      StackUpdate update = Builders.stackUpdate().template(template).files(subtemplate).parameters(params).build()
-      getRegionClient(region).heat().stacks().update(stackName, stackId, update)
-    }
-  }
+  @Delegate
+  OpenstackComputeV2Provider computeProvider
```
Comment (varikin) [inline clouddriver-openstack/src/main/groovy/com/netflix/spinnaker/clouddriver/openstack/client/OpenstackClientProvider.groovy:28]: V2 provider?

File: clouddriver-openstack/src/main/groovy/com/netflix/spinnaker/clouddriver/openstack/client/OpenstackProviderFactory.groovy [hunk 2]
```diff
@@ -16,42 +16,20 @@
...
-        .endpoint(credentials.endpoint)
-        .credentials(credentials.username, credentials.password, Identifier.byName(credentials.domainName))
-        .scopeToProject(Identifier.byName(credentials.tenantName), Identifier.byName(credentials.domainName))
-        .authenticate()
-      provider = new OpenstackClientV3Provider(osClient)
-    } else {
-      throw new IllegalArgumentException("Unknown account type ${credentials.accountType}")
-    }
-    provider
+    OpenstackIdentityProvider identityProvider = new OpenstackIdentityV3Provider(credentials)
```
Comment (varikin) [inline clouddriver-openstack/src/main/groovy/com/netflix/spinnaker/clouddriver/openstack/client/OpenstackProviderFactory.groovy:27]: Why always creating identity v3 provider?

File: clouddriver-openstack/src/main/groovy/com/netflix/spinnaker/clouddriver/openstack/client/OpenstackClientProvider.groovy (hunk 1 above)
Comment (drmaas) [inline clouddriver-openstack/src/main/groovy/com/netflix/spinnaker/clouddriver/openstack/client/OpenstackClientProvider.groovy:28]: Thats the version of compute - it's actually 2.0 or 2.1.

File: clouddriver-openstack/src/main/groovy/com/netflix/spinnaker/clouddriver/openstack/client/OpenstackProviderFactory.groovy (hunk 2 above)
Comment (drmaas) [inline clouddriver-openstack/src/main/groovy/com/netflix/spinnaker/clouddriver/openstack/client/OpenstackProviderFactory.groovy:27]: That's the version we are supporting, as it ships with liberty and is the current version. In the future we may upgrade to v4, if it comes out. The name just makes it explicit as to which version we are using.

Comment (varikin) [inline clouddriver-openstack/src/main/groovy/com/netflix/spinnaker/clouddriver/openstack/client/OpenstackProviderFactory.groovy:27]: Don't we use v2?

File: clouddriver-openstack/src/main/groovy/com/netflix/spinnaker/clouddriver/openstack/client/OpenstackClientProvider.groovy (hunk 1 above)
Comment (varikin) [inline clouddriver-openstack/src/main/groovy/com/netflix/spinnaker/clouddriver/openstack/client/OpenstackClientProvider.groovy:28]: Should this be a delegate to an interface without a version in it?

**Comment by drmaas** [comment]
reopening against spinnaker

**Pull request closed without merging**